In [8]:
import pandas as pd

In [10]:
months = pd.Series(range(1, 13))
months

0      1
1      2
2      3
3      4
4      5
5      6
6      7
7      8
8      9
9     10
10    11
11    12
dtype: int64

Có nhiều cách tính quantile, loại mặc định trong Pandas là linear, trong khi loại mình hay học và quen thuộc hơn là midpoint.

In [12]:
months.quantile([0.25, 0.5, 0.75], interpolation='midpoint') # tính 25%, 50%, 75% percentile

0.25    3.5
0.50    6.5
0.75    9.5
dtype: float64

In [13]:
months.quantile([0.25, 0.5, 0.75], interpolation='linear')

0.25    3.75
0.50    6.50
0.75    9.25
dtype: float64

Đầu tiên ta đếm số các điểm dữ liệu `n`, trong trường hợp này là 12. Để tính được `x` percentile, trước hết dùng công thức $ x*(n-1)+1 $ để tìm được index của điểm dữ liệu đầu tiên trong khoảng `x` percentile này. Với 25% percentile, $n=12, x=0.25$, ta sẽ bắt đầu từ "vị trí" $0.25*(12-1)+1= 3.75$.

Vị trí 3.75 ngụ ý điểm dữ liệu biểu diễn 25% percentile nằm ở giữa vị trí 3 và 4, cũng ứng với giá trị 3 và 4.

Với cách tính midpoint, ta lấy trung bình cộng của 2 giá trị này, nên 25% percentile sẽ là $\frac{3+4}{2}=3.5$.

Với cách tính linear, ta không lấy trung bình cộng, mà lấy giá trị $(3.75 - 3) * (4-3) + 3 = 3.75$. Đại loại là lấy phần lẻ của 3.75, nhân với khoảng cách giữa 2 giá trị 3 và 4, và cộng với 3. Ví dụ các số min_temperature sẽ dễ hiểu hơn.

In [14]:
min_temperature = pd.Series([57, 59, 61, 65, 66, 72, 72, 75, 76, 77, 77, 78])

In [15]:
min_temperature.quantile([0.25, 0.5, 0.75], interpolation='midpoint')

0.25    63.0
0.50    72.0
0.75    76.5
dtype: float64

In [16]:
min_temperature.quantile([0.25, 0.5, 0.75], interpolation='linear')

0.25    64.00
0.50    72.00
0.75    76.25
dtype: float64

Do số điểm dữ liệu cũng là 12, điểm dữ liệu biểu diễn 25% percentile cũng nằm ở giữa vị trí 3 và 4, lúc này là giá trị 61 và 65.
Với cách tính midpoint, 25% percentile là trung bình cộng $\frac{61+65}{2}=63.0$.
Với cách tính linear, 25% percentile là $0.75 * (65-61) + 61 = 64.0$.